fsspec/implementations/arrow.py

```diff
@@ -0,0 +1,241 @@
+import errno
+import io
+import os
+import secrets
+import shutil
+from contextlib import suppress
+from functools import wraps
+
+from fsspec.spec import AbstractFileSystem
+from fsspec.utils import infer_storage_options, mirror_from, stringify_path
+
+
+def wrap_exceptions(func):
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except OSError as exception:
+            if not exception.args:
+                raise
+
+            message, *args = exception.args
+            if isinstance(message, str) and "does not exist" in message:
+                raise FileNotFoundError(errno.ENOENT, message) from exception
+            else:
+                raise
+
+    return wrapper
```

Comment:
For the record: https://github.com/intake/filesystem_spec/pull/749#discussion_r706487444

Reply:
Unfortunately they are still present, without this many places I see just raise bare OSError s

Reply:
Feel free to open a JIRA on the Arrow side about this, if you still run into example cases that raise bare OSError for such cases.

Reply:
Here is a simple example, if I can get some time I will try to create a ticker for this and other scenarios I've experienced.
```terminal
(.venv38) (Python 3.8.5+) [  5:51ÖS ]  [ <hey>@<hey>:~/filesystem_spec(pyarrow✗) ]
 $ pip show pyarrow | grep -i version
Version: 4.0.0
License: Apache License, Version 2.0
(.venv38) (Python 3.8.5+) [  5:51ÖS ]  [ <hey>@<hey>:~/filesystem_spec(pyarrow✗) ]
 $ pytest fsspec/implementations/tests/test_arrow.py
============================================================================== test session starts ===============================================================================
platform linux -- Python 3.8.5+, pytest-6.2.5, py-1.9.0, pluggy-0.13.1
rootdir: /home/isidentical/filesystem_spec
plugins: timeout-1.4.2, mock-3.6.1, lazy-fixture-0.6.3, anyio-2.0.2, tap-3.2, env-0.6.2, cov-2.12.1, flaky-3.7.0, xdist-2.3.0, docker-0.10.3, apiritif-0.9.8, forked-1.3.0, asyncio-0.14.0
collected 11 items                                                                                                                                                               

fsspec/implementations/tests/test_arrow.py ........F..                                                                                                                     [100%]

==================================================================================== FAILURES ====================================================================================
________________________________________________________________________________ test_exceptions _________________________________________________________________________________

fs = <fsspec.implementations.arrow.ArrowFSWrapper object at 0x7f860c1810a0>, remote_dir = 'f41760ad91b75cc4ace4e99a94c7e9a8'

    def test_exceptions(fs, remote_dir):
        with pytest.raises(FileNotFoundError):
>           with fs.open(remote_dir + "/a.txt"):

fsspec/implementations/tests/test_arrow.py:156: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
fsspec/spec.py:978: in open
    f = self._open(
fsspec/implementations/arrow.py:144: in _open
    stream = self.fs.open_input_stream(path)
pyarrow/_fs.pyx:621: in pyarrow._fs.FileSystem.open_input_stream
    ???
pyarrow/error.pxi:141: in pyarrow.lib.pyarrow_internal_check_status
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   OSError: Path does not exist 'f41760ad91b75cc4ace4e99a94c7e9a8/a.txt'

pyarrow/error.pxi:112: OSError
============================================================================ short test summary info =============================================================================
FAILED fsspec/implementations/tests/test_arrow.py::test_exceptions - OSError: Path does not exist 'f41760ad91b75cc4ace4e99a94c7e9a8/a.txt'
========================================================================== 1 failed, 10 passed in 0.21s ==========================================================================
```